lib/msf/core/auxiliary/masscan.rb

```diff
@@ -0,0 +1,249 @@
+# -*- coding: binary -*-
+require 'rex/parser/masscan_nokogiri'
+require 'rex/parser/masscan_xml'
+require 'open3'
+
+module Msf
+
+###
+#
+# This module provides methods for interacting with masscan.
+# Modules that include this should define their own masscan_build_args()
+# function, and usually should have some method for dealing with
+# the data yielded from masscan_hosts(). See auxiliary/scanner/oracle/oracle_login
+# for an example implementation.
+#
+###
+
+module Auxiliary::Masscan
+
+attr_accessor :masscan_args, :masscan_bin, :masscan_log
+attr_reader :masscan_pid, :masscan_ver
+
+def initialize(info = {})
+  super
+
+  register_options([
+    OptAddressRange.new('RHOSTS', [ true, "The target address range or CIDR identifier"]),
+    OptBool.new('MASSCAN_VERBOSE', [ false, 'Display masscan output', true]),
+    OptString.new('RPORTS', [ false, 'Ports to target']), # RPORT supersedes RPORTS
+  ], Auxiliary::Masscan)
+
+end
+
+def setup
+  print "setup"
+  deregister_options("RPORT")
+  @masscan_args = []
+  @masscan_bin = masscan_binary_path
+  # set masscan cmd
+  self.masscan_bin || (raise RuntimeError, "Cannot locate masscan binary")
+  masscan_set_log
+  masscan_add_ports
+  masscan_cmd = [self.masscan_bin]
+  self.masscan_args.unshift("-oX #{self.masscan_log[1]}")
+  masscan_cmd << self.masscan_args.join(" ")
+  masscan_cmd << datastore['RHOSTS']
+  masscan_cmd.join(" ")
+end
+
+def rports
+    datastore['RPORTS']
+end
+
+def rport
+    datastore['RPORT']
+end
+
+def masscan_build_args
+  raise RuntimeError, "masscan_build_args() not defined by #{self.refname}"
```

Comment:
Here and most other places this should be a `fail_with`, which I'll fix in a PR for you.